### PR TITLE
Restructured EventHub to new child-resources structure

### PR DESCRIPTION
**Pull request by MrMCake**
# Change

- Restructured EventHub to new child-resources structure
- Moved scripts into shared utilities (from which they can also be used by the later ADO pipelines)

Pipeline reference
[EventHub: Namespaces](https://github.com/Azure/ResourceModules/actions/workflows/ms.eventhub.namespaces.yml)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: arm/Microsoft.EventHub/namespaces/deploy.bicep [hunk 1]
```diff
@@ -188,31 +184,69 @@ resource eventHubNamespace_lock 'Microsoft.Authorization/locks@2016-09-01' = if
...
   name: '${eventHubNamespace.name}-diagnosticSettings'
   properties: {
-    storageAccountId: (empty(diagnosticStorageAccountId) ? json('null') : diagnosticStorageAccountId)
-    workspaceId: (empty(workspaceId) ? json('null') : workspaceId)
-    metrics: ((empty(diagnosticStorageAccountId) && empty(workspaceId)) ? json('null') : diagnosticsMetrics)
-    logs: ((empty(diagnosticStorageAccountId) && empty(workspaceId)) ? json('null') : diagnosticsLogs)
+    storageAccountId: (empty(diagnosticStorageAccountId) ? null : diagnosticStorageAccountId)
+    workspaceId: (empty(workspaceId) ? null : workspaceId)
+    metrics: ((empty(diagnosticStorageAccountId) && empty(workspaceId)) ? null : diagnosticsMetrics)
+    logs: ((empty(diagnosticStorageAccountId) && empty(workspaceId)) ? null : diagnosticsLogs)
```
Comment (MariusStorhaug) [inline arm/Microsoft.EventHub/namespaces/deploy.bicep:190]: Clean up '(' and ')' pairs and align on 
```
condition ? parameter-value : default-value
```
```suggestion
    storageAccountId: !empty(diagnosticStorageAccountId) ? diagnosticStorageAccountId : null
    workspaceId: !empty(workspaceId) ? workspaceId : null
    metrics: !empty(diagnosticStorageAccountId) || !empty(workspaceId) ? diagnosticsMetrics : null
    logs: !empty(diagnosticStorageAccountId) || !empty(workspaceId)) ? diagnosticsLogs : null
```

File: arm/Microsoft.EventHub/namespaces/deploy.bicep [hunk 2]
```diff
@@ -172,7 +168,7 @@ resource eventHubNamespace 'Microsoft.EventHub/namespaces@2017-04-01' = {
     zoneRedundant: zoneRedundant
     isAutoInflateEnabled: isAutoInflateEnabled
     maximumThroughputUnits: maximumThroughputUnits_var
-    networkAcls: (empty(networkAcls) ? json('null') : networkAcls_var)
+    networkAcls: (empty(networkAcls) ? null : networkAcls_var)
```
Comment (MariusStorhaug) [inline arm/Microsoft.EventHub/namespaces/deploy.bicep:171]:
```suggestion
    networkAcls: !empty(networkAcls) ? networkAcls_var : null
```

File: arm/Microsoft.EventHub/namespaces/eventhubs/deploy.bicep [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+resource eventHub 'Microsoft.EventHub/namespaces/eventhubs@2021-06-01-preview' = {
+  name: '${namespaceName}/${name}'
+  properties: captureDescriptionEnabled ? eventHubPropertiesWithCapture : eventHubPropertiesSimple
+}
+
+resource eventHub_lock 'Microsoft.Authorization/locks@2016-09-01' = if (lock != 'NotSpecified') {
+  name: '${last(split(eventHub.name, '/'))}-${lock}-lock'
+  properties: {
+    level: lock
+    notes: (lock == 'CanNotDelete') ? 'Cannot delete resource or child resources.' : 'Cannot modify the resource or child resources.'
```
Comment (MariusStorhaug) [inline arm/Microsoft.EventHub/namespaces/eventhubs/deploy.bicep:139]:
```suggestion
    notes: lock == 'CanNotDelete' ? 'Cannot delete resource or child resources.' : 'Cannot modify the resource or child resources.'
```

File: arm/Microsoft.EventHub/namespaces/eventhubs/.bicep/nested_rbac.bicep [hunk 4]
```diff
@@ -25,10 +25,9 @@ resource eventHub 'Microsoft.EventHub/namespaces/eventhubs@2021-06-01-preview' e
 }
 
 resource roleAssignment 'Microsoft.Authorization/roleAssignments@2020-04-01-preview' = [for principalId in roleAssignmentObj.principalIds: {
-  name: guid(eventHub.name, principalId, roleAssignmentObj.roleDefinitionIdOrName)
+  name: guid(resourceName, principalId, roleAssignmentObj.roleDefinitionIdOrName)
   properties: {
     roleDefinitionId: contains(builtInRoleNames, roleAssignmentObj.roleDefinitionIdOrName) ? builtInRoleNames[roleAssignmentObj.roleDefinitionIdOrName] : roleAssignmentObj.roleDefinitionIdOrName
     principalId: principalId
   }
-  scope: eventHub
```
Comment (MariusStorhaug) [inline arm/Microsoft.EventHub/namespaces/eventhubs/.bicep/nested_rbac.bicep:33]: Why are you removing this? If you remove this the scope will be the deployment scope of RG i think.

File: arm/Microsoft.EventHub/namespaces/deploy.bicep [hunk 5]
```diff
@@ -88,24 +80,28 @@ param cuaId string = ''
...
 }]
 var networkAcls_var = {
-  bypass: (empty(networkAcls) ? json('null') : networkAcls.bypass)
-  defaultAction: (empty(networkAcls) ? json('null') : networkAcls.defaultAction)
-  virtualNetworkRules: (empty(networkAcls) ? json('null') : virtualNetworkRules)
-  ipRules: (empty(networkAcls) ? json('null') : ((length(networkAcls.ipRules) == 0) ? json('null') : networkAcls.ipRules))
+  bypass: empty(networkAcls) ? null : networkAcls.bypass
+  defaultAction: empty(networkAcls) ? null : networkAcls.defaultAction
+  virtualNetworkRules: empty(networkAcls) ? null : virtualNetworkRules
+  ipRules: empty(networkAcls) ? null : ((length(networkAcls.ipRules) == 0) ? null : networkAcls.ipRules)
```
Comment (MariusStorhaug) [inline arm/Microsoft.EventHub/namespaces/deploy.bicep:100]:
```suggestion
  bypass: !empty(networkAcls) ? networkAcls.bypass : null
  defaultAction: !empty(networkAcls) ? networkAcls.defaultAction : null
  virtualNetworkRules: !empty(networkAcls) ? virtualNetworkRules : null
  ipRules: !empty(networkAcls) ? (length(networkAcls.ipRules) > 0 ? networkAcls.ipRules : null) : null
```

File: arm/Microsoft.EventHub/namespaces/deploy.bicep [hunk 6]
```diff
@@ -88,24 +80,28 @@ param cuaId string = ''
...
 }
-var namespaceAlias_var = (empty(namespaceAlias) ? 'placeholder' : namespaceAlias)
+
+@description('Optional. The disaster recovery config for this namespace')
+param disasterRecoveryConfig object = {}
```
Comment (MariusStorhaug) [inline arm/Microsoft.EventHub/namespaces/deploy.bicep:104]: Should we try to keep the bicep file structured to have:
1. params
1. vars
1. resources/modules
1. outputs

In other words, move params over all the var declarations

File: arm/Microsoft.EventHub/namespaces/.bicep/nested_privateEndpoint.bicep [hunk 7]
```diff
@@ -11,7 +11,7 @@ var privateEndpoint_var = {
     privateEndpointObj.service
   ]
   privateDnsZoneResourceIds: (contains(privateEndpointObj, 'privateDnsZoneResourceIds') ? (empty(privateEndpointObj.privateDnsZoneResourceIds) ? [] : privateEndpointObj.privateDnsZoneResourceIds) : [])
-  customDnsConfigs: (contains(privateEndpointObj, 'customDnsConfigs') ? (empty(privateEndpointObj.customDnsConfigs) ? json('null') : privateEndpointObj.customDnsConfigs) : json('null'))
+  customDnsConfigs: (contains(privateEndpointObj, 'customDnsConfigs') ? (empty(privateEndpointObj.customDnsConfigs) ? null : privateEndpointObj.customDnsConfigs) : null)
```
Comment (MariusStorhaug) [inline arm/Microsoft.EventHub/namespaces/.bicep/nested_privateEndpoint.bicep:14]:
```suggestion
  customDnsConfigs: contains(privateEndpointObj, 'customDnsConfigs') ? (!empty(privateEndpointObj.customDnsConfigs) ? privateEndpointObj.customDnsConfigs : null) : null
```

**Review by MariusStorhaug — CHANGES_REQUESTED** [review]

**Comment by MariusStorhaug** [comment]
Suggestion: Create separate PRs for separate topics. Now the script fix is being blocked by changes requested on evenhub topics.

File: arm/Microsoft.EventHub/namespaces/eventhubs/.bicep/nested_rbac.bicep (hunk 4 above)
Comment (MrMCake) [reply arm/Microsoft.EventHub/namespaces/eventhubs/.bicep/nested_rbac.bicep]: No idea. Added it again

File: arm/Microsoft.EventHub/namespaces/deploy.bicep (hunk 6 above)
Comment (MrMCake) [reply arm/Microsoft.EventHub/namespaces/deploy.bicep:104]: yes we should

Comment (MrMCake) [reply arm/Microsoft.EventHub/namespaces/deploy.bicep:87]: done

**Review by MariusStorhaug — APPROVED** [review]
Conversations resolved, fixes done. Approved.

**Pull request merged**